### ISSUE TITLE
grpc.Errorf should return struct pointer

### ISSUE THREAD

**Issue by danruehle** [grpc/grpc-go#698]
Using `go version go1.6.2 darwin/amd64` on Mac OSX.

Can define multiple errors via grpc.Errorf and two of these have the same code and string value. Those two errors will be considered equal, so returning either one will look like returning either one. This occurs because grpc.Errorf returns struct, not struct pointer. The fmt.Errorf function returns a struct pointer so it doesn't suffer from this issue because when two interface variables that contain pointers are compared, they are equal only if they point to the same thing. If the two interfaces contain structs, the structs themselves are compared per the golang doc 

> Interface values are comparable. Two interface values are equal if they have identical dynamic types and equal dynamic values or if both have value nil.

It seems like grpc.Errorf should return a struct pointer to keep the error comparison semantic working.

Code to reproduce:

```golang
package _playground

import (
	"fmt"
	"testing"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
)

var (
	// ErrStandard is a standard error
	ErrStandard = fmt.Errorf("This is a standard error")
	// ErrStandardDup is a standard error with the same message as ErrStandard
	ErrStandardDup = fmt.Errorf("This is a standard error")

	// ErrGrpcStandard is a standard GRPC error
	ErrGrpcStandard = grpc.Errorf(codes.InvalidArgument, "This is a standard error")
	//ErrGrpcStandardDup is a standard GRPC error with the same code and message as ErrGrpcStandard
	ErrGrpcStandardDup = grpc.Errorf(codes.InvalidArgument, "This is a standard error")
)

func TestFmt(t *testing.T) {
	if ErrStandard == ErrStandardDup {
		t.Fatalf("fmt.Errorf instances are considered equal: %v - %v", ErrStandard, ErrStandardDup)
	}
}

func TestGrpc(t *testing.T) {
	if ErrGrpcStandard == ErrGrpcStandardDup {
		t.Fatalf("grpc.Errorf instances are considered equal: %v - %v", ErrGrpcStandard, ErrGrpcStandardDup)
	}
}
```

**Comment by iamqizhao** [comment]
sounds good. Feel free to contribute.

**Comment by danruehle** [comment]
Working on it. There are several tests that use the comparing of error interfaces being equal. I question whether this could possibly be a breaking change for users of the package?

**Comment by iamqizhao** [comment]
closed via #755
